File: code/packages/python/virtual-memory/tests/test_page_table_entry.py
"""Tests for PageTableEntry — the metadata for a single page mapping."""

from virtual_memory.page_table_entry import (
    PAGE_OFFSET_BITS,
    PAGE_SIZE,
    VPN_BITS,
    PageTableEntry,
)


class TestConstants:
    """Verify the fundamental page geometry constants."""

    def test_page_size_is_4kb(self) -> None:
        """4 KB pages are the standard since the Intel 386 (1985)."""
        assert PAGE_SIZE == 4096

    def test_page_offset_bits_is_12(self) -> None:
        """12 bits address every byte within a 4 KB page (2^12 = 4096)."""
        assert PAGE_OFFSET_BITS == 12

    def test_vpn_bits_is_20(self) -> None:
        """20 bits for VPN: 2^20 pages * 4 KB = 4 GB address space."""
        assert VPN_BITS == 20

    def test_page_size_matches_offset_bits(self) -> None:
        """The page size must be 2^offset_bits."""
        assert PAGE_SIZE == (1 << PAGE_OFFSET_BITS)

    def test_total_address_bits(self) -> None:
        """VPN bits + offset bits = 32 (full 32-bit address space)."""
        assert VPN_BITS + PAGE_OFFSET_BITS == 32


class TestPageTableEntry:
    """Test PTE creation and flag behavior."""

    def test_default_values(self) -> None:
        """A fresh PTE has sensible defaults: not present, frame 0."""
        pte = PageTableEntry()
        assert pte.frame_number == 0
        assert pte.present is False
        assert pte.dirty is False
        assert pte.accessed is False
        assert pte.writable is True
        assert pte.executable is False
        assert pte.user_accessible is True

    def test_custom_frame_number(self) -> None:
        """Frame number can be set at creation."""
        pte = PageTableEntry(frame_number=42)
        assert pte.frame_number == 42

    def test_present_flag(self) -> None:
        """The present flag indicates the page is in physical memory."""
        pte = PageTableEntry(present=True)
        assert pte.present is True

    def test_dirty_flag(self) -> None:
        """The dirty flag indicates the page has been written to."""
        pte = PageTableEntry(dirty=True)
        assert pte.dirty is True

    def test_accessed_flag(self) -> None:
        """The accessed flag indicates the page was recently used."""
        pte = PageTableEntry(accessed=True)
        assert pte.accessed is True

    def test_writable_flag(self) -> None:
        """Pages can be made read-only (e.g., code pages)."""
        pte = PageTableEntry(writable=False)
        assert pte.writable is False

    def test_executable_flag(self) -> None:
        """The executable flag controls the NX (no-execute) bit."""
        pte = PageTableEntry(executable=True)
        assert pte.executable is True

    def test_user_accessible_flag(self) -> None:
        """Kernel pages are not user-accessible."""
        pte = PageTableEntry(user_accessible=False)
        assert pte.user_accessible is False

    def test_all_flags_set(self) -> None:
        """All flags can be set simultaneously."""
        pte = PageTableEntry(
            frame_number=100,
            present=True,
            dirty=True,
            accessed=True,
            writable=True,
            executable=True,
            user_accessible=True,
        )
        assert pte.frame_number == 100
        assert pte.present is True
        assert pte.dirty is True
        assert pte.accessed is True
        assert pte.writable is True
        assert pte.executable is True
        assert pte.user_accessible is True

    def test_flags_are_mutable(self) -> None:
        """PTE flags can be changed after creation (hardware does this)."""
        pte = PageTableEntry()
        pte.accessed = True
        pte.dirty = True
        assert pte.accessed is True
        assert pte.dirty is True

    def test_copy_creates_independent_instance(self) -> None:
        """copy() creates a deep copy — modifying one doesn't affect the other."""
        original = PageTableEntry(frame_number=5, present=True, writable=True)
        copy = original.copy()

        # Same values initially.
        assert copy.frame_number == 5
        assert copy.present is True
        assert copy.writable is True

        # Modify the copy — original should be unaffected.
        copy.frame_number = 99
        copy.writable = False
        assert original.frame_number == 5
        assert original.writable is True

    def test_copy_preserves_all_flags(self) -> None:
        """copy() preserves every single flag."""
        original = PageTableEntry(
            frame_number=42,
            present=True,
            dirty=True,
            accessed=True,
            writable=False,
            executable=True,
            user_accessible=False,
        )
        copy = original.copy()
        assert copy.frame_number == 42
        assert copy.present is True
        assert copy.dirty is True
        assert copy.accessed is True
        assert copy.writable is False
        assert copy.executable is True
        assert copy.user_accessible is False

File: code/packages/ruby/virtual_memory/test/test_page_table_entry.rb
# frozen_string_literal: true

require_relative "test_helper"

module CodingAdventures
  module VirtualMemory
    class TestPageTableEntry < Minitest::Test
      # == Default Values ==
      #
      # A freshly created PTE should have sensible defaults:
      # not present (no physical frame assigned yet), not dirty,
      # not accessed, writable (most pages are), not executable,
      # and user-accessible.

      def test_default_values
        pte = PageTableEntry.new

        assert_equal 0, pte.frame_number
        refute pte.present?
        refute pte.dirty?
        refute pte.accessed?
        assert pte.writable?
        refute pte.executable?
        assert pte.user_accessible?
      end

      # == Custom Initialization ==
      #
      # All flags can be set at creation time. This is used when
      # creating a PTE for a known mapping (e.g., mapping a code
      # page that should be executable but read-only).

      def test_custom_initialization
        pte = PageTableEntry.new(
          frame_number: 42,
          present: true,
          dirty: true,
          accessed: true,
          writable: false,
          executable: true,
          user_accessible: false
        )

        assert_equal 42, pte.frame_number
        assert pte.present?
        assert pte.dirty?
        assert pte.accessed?
        refute pte.writable?
        assert pte.executable?
        refute pte.user_accessible?
      end

      # == Flag Mutation ==
      #
      # The hardware (simulated) sets the accessed and dirty bits
      # during memory access. The OS clears them during page
      # replacement scanning.

      def test_flag_mutation
        pte = PageTableEntry.new(frame_number: 10, present: true)

        # Simulate a read access.
        pte.accessed = true
        assert pte.accessed?

        # Simulate a write access.
        pte.dirty = true
        assert pte.dirty?

        # OS clears accessed bit during clock sweep.
        pte.accessed = false
        refute pte.accessed?
      end

      # == Frame Number ==
      #
      # The frame number can be changed when a page is remapped
      # (e.g., after a COW fault allocates a new frame).

      def test_frame_number_update
        pte = PageTableEntry.new(frame_number: 5, present: true)
        assert_equal 5, pte.frame_number

        pte.frame_number = 99
        assert_equal 99, pte.frame_number
      end

      # == Duplication ==
      #
      # PTE duplication is used during fork() to create independent
      # copies of page table entries.

      def test_dup_creates_independent_copy
        original = PageTableEntry.new(
          frame_number: 7,
          present: true,
          dirty: true,
          writable: true
        )

        copy = original.dup

        # Copy should have the same values.
        assert_equal 7, copy.frame_number
        assert copy.present?
        assert copy.dirty?
        assert copy.writable?

        # Modifying the copy should not affect the original.
        copy.frame_number = 99
        copy.writable = false

        assert_equal 7, original.frame_number
        assert original.writable?
      end

      # == Permission Combinations ==
      #
      # Real page table entries use combinations of permission bits
      # to represent different page types:

      def test_code_page_permissions
        # Code pages: readable + executable, NOT writable.
        # This prevents code injection attacks.
        pte = PageTableEntry.new(
          frame_number: 1,
          present: true,
          writable: false,
          executable: true,
          user_accessible: true
        )

        refute pte.writable?
        assert pte.executable?
        assert pte.user_accessible?
      end

      def test_data_page_permissions
        # Data pages (stack/heap): readable + writable, NOT executable.
        # This is the NX (No-Execute) bit in action.
        pte = PageTableEntry.new(
          frame_number: 2,
          present: true,
          writable: true,
          executable: false,
          user_accessible: true
        )

        assert pte.writable?
        refute pte.executable?
      end

      def test_kernel_page_permissions
        # Kernel pages: not user-accessible.
        # User-mode code accessing a kernel page triggers a fault.
        pte = PageTableEntry.new(
          frame_number: 3,
          present: true,
          writable: true,
          executable: true,
          user_accessible: false
        )

        refute pte.user_accessible?
      end

      # == Present Bit ==
      #
      # A page that is not present triggers a page fault on access.
      # This is the basis for demand paging and swapping.

      def test_not_present_page
        pte = PageTableEntry.new(frame_number: 0, present: false)
        refute pte.present?

        # Making it present (after fault handler allocates a frame).
        pte.present = true
        pte.frame_number = 42
        assert pte.present?
        assert_equal 42, pte.frame_number
      end
    end
  end
end
